crates/ruff_python_ast/src/source_code/stylist.rs

```diff
@@ -14,42 +15,82 @@ use crate::str::leading_quote;
 use crate::types::Range;
 
 pub struct Stylist<'a> {
-    contents: &'a str,
     locator: &'a Locator<'a>,
     indentation: OnceCell<Indentation>,
+    indent_end: Option<Location>,
     quote: OnceCell<Quote>,
+    quote_range: Option<Range>,
     line_ending: OnceCell<LineEnding>,
 }
 
 impl<'a> Stylist<'a> {
     pub fn indentation(&'a self) -> &'a Indentation {
-        self.indentation
-            .get_or_init(|| detect_indentation(self.contents, self.locator).unwrap_or_default())
+        self.indentation.get_or_init(|| {
+            if let Some(indent_end) = self.indent_end {
+                let start = Location::new(indent_end.row(), 0);
+                let whitespace = self.locator.slice(Range::new(start, indent_end));
+                Indentation(whitespace.to_string())
+            } else {
+                Indentation::default()
+            }
+        })
     }
 
-    pub fn quote(&'a self) -> &'a Quote {
-        self.quote
-            .get_or_init(|| detect_quote(self.contents, self.locator).unwrap_or_default())
+    pub fn quote(&'a self) -> Quote {
+        *self.quote.get_or_init(|| {
+            self.quote_range
+                .and_then(|quote_range| {
+                    let content = self.locator.slice(quote_range);
+                    leading_quote(&content)
+                })
+                .map(|pattern| {
+                    if pattern.contains('\'') {
+                        Quote::Single
+                    } else if pattern.contains('"') {
+                        Quote::Double
+                    } else {
+                        unreachable!("Expected string to start with a valid quote prefix")
+                    }
+                })
+                .unwrap_or_default()
+        })
     }
 
     pub fn line_ending(&'a self) -> &'a LineEnding {
         self.line_ending
-            .get_or_init(|| detect_line_ending(self.contents).unwrap_or_default())
+            .get_or_init(|| detect_line_ending(self.locator.contents()).unwrap_or_default())
     }
 
-    pub fn from_contents(contents: &'a str, locator: &'a Locator<'a>) -> Self {
+    pub fn from_tokens(tokens: &[LexResult], locator: &'a Locator<'a>) -> Self {
+        let indent_end = tokens.iter().flatten().find_map(|(_, t, end)| {
+            if matches!(t, Tok::Indent) {
+                Some(*end)
+            } else {
+                None
+            }
+        });
+
+        let quote_range = tokens.iter().flatten().find_map(|(start, t, end)| match t {
+            Tok::String {
+                triple_quoted: false,
+                ..
+            } => Some(Range::new(*start, *end)),
+            _ => None,
+        });
```

Comment:
Ideally, we wouldn't need to iterate here but that's not possible without cloning `tokens` because `parse_program` requires owned tokens. 

Iterating over some tokens here should be significantly cheaper than re-lexing all the tokens.

Reply:
If we're doing this already, would it make sense to remove the `OnceCell` entirely from `indentation` and `quote`?

Reply:
Not sure. The problem is that we'll use `Locator` which I, as far I understand, we should avoid because building the index is expensive

Reply:
Eh yeah, alright, let's leave it as-is. At some point we should get some data or otherwise develop intuition around how often `Locator` is actually getting initialized (i.e., is every file populating it anyway, despite doing so lazily? Or no?).